extensions/amp-analytics/0.1/amp-analytics.js

```diff
@@ -331,12 +335,63 @@ export class AmpAnalytics extends AMP.BaseElement {
       request = addParamsToUrl(request, this.config_['extraUrlParams']);
     }
 
-    this.config_['vars']['requestCount']++;
 
+    return this.isSampledIn_(trigger).then(result => {
+      if (!result) {
+        return Promise.resolve('');
+      }
+
+      this.config_['vars']['requestCount']++;
+      request = this.expandTemplate_(request, trigger);
+
+      // For consistency with amp-pixel we also expand any url replacements.
+      return urlReplacementsFor(this.getWin()).expand(request).then(request => {
+        this.sendRequest_(request, trigger);
+        return request;
+      });
+    });
+  }
+
+  /**
+   * @param {!JSONObject} trigger The config to use to determine sampling.
+   * @return {!Promise.<boolean>} Whether the request should be sampled in or
+   * not based on sampleSpec.
+   * @private
+   */
+  isSampledIn_(trigger) {
+    const spec = trigger['sampleSpec'];
+    const resolve = Promise.resolve(true);
+    if (!spec) {
+      return resolve;
+    }
+    const threshold = spec['threshold'];
+    if (!spec['sampleOn'] ||
+        Number.isNaN(parseFloat(threshold)) || !Number.isFinite(threshold)) {
+      console./*OK*/error(this.getName_(), 'Invalid sampling spec.');
+      return resolve;
+    }
+    const key = this.expandTemplate_(spec['sampleOn'], trigger);
```

Comment:
`spec['sampleOn']` can be any string with one or more macros. I used `requestCount` in test but it could be `RANDOM` or `CLIENT_ID(...)` in real world.

`this.expandTemplate_(spec['sampleOn'], trigger)` uses var definitions from `vendors.js` and  trigger configs and converts them into something that can be used by `url-replacements.js` class. The call to `urlReplacementsFor(this.getWin()).expand()` below will then convert that template string into another string with macros substituted for real values.

In the tests, the value of requestCount will always be 0 since no requests is generated at that point. But that actually doesn't matter because I stub the values returned by sha384. I can change the tests to use `RANDOM` instead if that makes it less confusing.